### PR TITLE
replace mutt_mktemp() with buf_mktemp()

### PR DESCRIPTION
**Pull request by roccoblues**
This replaces `mutt_mktemp()` with the Buffer version `buf_mktemp()`, as discussed on the mailing list.

~I've only changed `send/send.c` so far because this is my first contribution and I wanted to get some early feedback. Since `mutt_mktemp()` isn't used in that many places I'd probably do all at once in this PR? Or do you prefer to have it split into smaller chunks?~

### CODE REVIEW

**Comment by flatcap** [comment]
> get some early feedback

The code looks good.

> I'd probably do all at once in this PR?

Great

> Or do you prefer to have it split into smaller chunks?

No need for these changes, as they're simple.

File: attach/recvattach.c [hunk 1]
```diff
@@ -553,28 +553,25 @@ static void query_pipe_attachment(const char *command, FILE *fp, struct Body *bo
       msgwin_clear_text();
       return;
     }
-    mutt_mktemp(tfile, sizeof(tfile));
-  }
-  else
-  {
-    tfile[0] = '\0';
```
Comment (roccoblues) [inline attach/recvattach.c:560]: I didn't understand why need to NULL terminate here. Please double check.

Comment (flatcap) [reply attach/recvattach.c:560]: You're right, this isn't needed any more.
It was made redundant by commit 7380cb5fae63481d5d8c10c9a669ebfafcd80ef3 (a year ago), which initialises `tfile`.

File: nntp/nntp.c [hunk 2]
```diff
@@ -2654,6 +2655,7 @@ static bool nntp_msg_open(struct Mailbox *m, struct Message *msg, struct Email *
     }
 
     /* fetch message to cache file */
+    char buf[PATH_MAX] = { 0 };
```
Comment (roccoblues) [inline nntp/nntp.c:2658]: `PATH_MAX` is a bit misleading now because the buffer is not used for a path anymore. What do you think?

Comment (flatcap) [reply nntp/nntp.c:2658]: true, but it still needs to be quite large for a message-id
```c
snprintf(buf, sizeof(buf), "ARTICLE %s\r\n", nntp_edata_get(e)->article_num ? article : e->env->message_id);
```
https://github.com/neomutt/neomutt/blob/07c62c30ad149c194440386bd8e43fb506413820/nntp/nntp.c#L2657C1-L2658

Comment (flatcap) [reply nntp/nntp.c:2658]: The RFC doesn't seem to limit the message-id, so 2048 bytes?
(and leave converting the rest of the nntp driver for another day)

Comment (roccoblues) [reply nntp/nntp.c]: 👍🏼 I've used 2048 bytes now.

File: attach/recvattach.c [hunk 3]
```diff
@@ -553,28 +553,25 @@ static void query_pipe_attachment(const char *command, FILE *fp, struct Body *bo
       msgwin_clear_text();
       return;
```
Comment (flatcap) [inline attach/recvattach.c:554]: If the user cancels the op, we leak `tfile` here

File: attach/recvattach.c [hunk 4]
```diff
@@ -553,28 +553,25 @@ static void query_pipe_attachment(const char *command, FILE *fp, struct Body *bo
...
-    mutt_mktemp(tfile, sizeof(tfile));
-  }
-  else
-  {
-    tfile[0] = '\0';
+    buf_mktemp(tfile);
   }
 
-  if (mutt_pipe_attachment(fp, body, command, tfile))
+  if (mutt_pipe_attachment(fp, body, command, buf_strdup(tfile)))
```
Comment (flatcap) [inline attach/recvattach.c:559]: We want `buf_string()` here, or we'll leak

File: color/command.c [hunk 5]
```diff
@@ -863,17 +863,17 @@ static enum CommandResult color_dump(struct Buffer *buf, struct Buffer *s,
...
-  fp_out = mutt_file_fopen(tempfile, "w");
+  buf_mktemp(tempfile);
+  fp_out = mutt_file_fopen(buf_string(tempfile), "w");
   if (!fp_out)
   {
     // L10N: '%s' is the file name of the temporary file
-    buf_printf(err, _("Could not create temporary file %s"), tempfile);
+    buf_printf(err, _("Could not create temporary file %s"), buf_string(tempfile));
     buf_dealloc(&filebuf);
     return MUTT_CMD_ERROR;
```
Comment (flatcap) [inline color/command.c:878]: We leak `tempfile` here

File: color/command.c [hunk 6]
```diff
@@ -1023,7 +1023,8 @@ static enum CommandResult color_dump(struct Buffer *buf, struct Buffer *s,
   struct PagerData pdata = { 0 };
   struct PagerView pview = { &pdata };
 
-  pdata.fname = tempfile;
+  pdata.fname = buf_strdup(tempfile);
+  buf_pool_release(&tempfile);
```
Comment (flatcap) [inline color/command.c:1027]: This is an odd bit of code...
We've created a couple of temporary structs (`pdata`, `pview`) to pass to `mutt_do_pager()`.
They don't want to take ownership of `tempfile`, they just share it: `buf_string()`.
That means we can't `buf_pool_release()` until after `mutt_do_pager()`.

File: commands.c [hunk 7]
```diff
@@ -735,14 +735,14 @@ enum CommandResult parse_my_hdr(struct Buffer *buf, struct Buffer *s,
...
+  buf_mktemp(tempfile);
 
-  FILE *fp_out = mutt_file_fopen(tempfile, "w");
+  FILE *fp_out = mutt_file_fopen(buf_string(tempfile), "w");
   if (!fp_out)
   {
     // L10N: '%s' is the file name of the temporary file
-    buf_printf(err, _("Could not create temporary file %s"), tempfile);
+    buf_printf(err, _("Could not create temporary file %s"), buf_string(tempfile));
     return MUTT_CMD_ERROR;
```
Comment (flatcap) [inline commands.c:746]: leak

File: commands.c [hunk 8]
```diff
@@ -753,7 +753,8 @@ enum CommandResult set_dump(ConfigDumpFlags flags, struct Buffer *err)
   struct PagerData pdata = { 0 };
   struct PagerView pview = { &pdata };
 
-  pdata.fname = tempfile;
+  pdata.fname = buf_strdup(tempfile);
+  buf_pool_release(&tempfile);
```
Comment (flatcap) [inline commands.c:757]: `buf_string()`, then `buf_pool_release()` after `mutt_do_pager()`

File: gui/global.c [hunk 9]
```diff
@@ -93,10 +93,10 @@ static int op_shell_escape(int op)
...
-  char tempfile[PATH_MAX] = { 0 };
-  mutt_mktemp(tempfile, sizeof(tempfile));
+  struct Buffer *tempfile = buf_pool_get();
+  buf_mktemp(tempfile);
 
-  FILE *fp = mutt_file_fopen(tempfile, "a+");
+  FILE *fp = mutt_file_fopen(buf_string(tempfile), "a+");
   if (!fp)
   {
     mutt_perror("fopen");
```
Comment (flatcap) [inline gui/global.c:102]: leak

File: gui/global.c [hunk 10]
```diff
@@ -109,7 +109,9 @@ static int op_show_log_messages(int op)
   struct PagerData pdata = { 0 };
   struct PagerView pview = { &pdata };
 
-  pdata.fname = tempfile;
+  pdata.fname = buf_strdup(tempfile);
+
+  buf_pool_release(&tempfile);
```
Comment (flatcap) [inline gui/global.c:114]: `buf_string()`, then `buf_pool_release()` after `mutt_do_pager()`

File: keymap.c [hunk 11]
```diff
@@ -1481,12 +1480,13 @@ static enum CommandResult dump_bind_macro(struct Buffer *buf, struct Buffer *s,
...
+  struct Buffer *tempfile = buf_pool_get();
+  buf_mktemp(tempfile);
+  fp_out = mutt_file_fopen(buf_string(tempfile), "w");
   if (!fp_out)
   {
     // L10N: '%s' is the file name of the temporary file
-    buf_printf(err, _("Could not create temporary file %s"), tempfile);
+    buf_printf(err, _("Could not create temporary file %s"), buf_string(tempfile));
     buf_dealloc(&filebuf);
     return MUTT_CMD_ERROR;
```
Comment (flatcap) [inline keymap.c:1491]: leak

File: keymap.c [hunk 12]
```diff
@@ -1498,7 +1498,8 @@ static enum CommandResult dump_bind_macro(struct Buffer *buf, struct Buffer *s,
   struct PagerData pdata = { 0 };
   struct PagerView pview = { &pdata };
 
-  pdata.fname = tempfile;
+  pdata.fname = buf_strdup(tempfile);
+  buf_pool_release(&tempfile);
```
Comment (flatcap) [inline keymap.c:1502]: `buf_string()`, then `buf_pool_release()` after `mutt_do_pager()`

File: send/send.c [hunk 13]
```diff
@@ -3011,11 +3012,11 @@ static bool send_simple_email(struct Mailbox *m, struct EmailArray *ea,
...
-  mutt_mktemp(tempfile, sizeof(tempfile));
+  struct Buffer *tempfile = buf_pool_get();
+  buf_mktemp(tempfile);
   if (body)
   {
-    FILE *fp = mutt_file_fopen(tempfile, "w+");
+    FILE *fp = mutt_file_fopen(buf_string(tempfile), "w+");
     if (!fp)
     {
       email_free(&e);
```
Comment (flatcap) [inline send/send.c:3022]: leak

**Review by flatcap — CHANGES_REQUESTED** [review]
It looks pretty good. so far.
I've made a lot of comments, but it's just a couple of simply-fixed bugs in multiple places.
Thanks!

File: attach/recvattach.c (hunk 4 above)
Comment (roccoblues) [reply attach/recvattach.c:560]: Is it ok if I change `mutt_pipe_attachment()`?

```c
-int mutt_pipe_attachment(FILE *fp, struct Body *b, const char *path, char *outfile);
+int mutt_pipe_attachment(FILE *fp, struct Body *b, const char *path, const char *outfile);
```

File: color/command.c (hunk 6 above)
Comment (roccoblues) [reply color/command.c]: good catch, thanks! Should be fixed now.

File: gui/global.c (hunk 10 above)
Comment (roccoblues) [reply gui/global.c]: should be fixed now

File: keymap.c (hunk 12 above)
Comment (roccoblues) [reply keymap.c]: should be fixed now

File: attach/recvattach.c (hunk 4 above)
Comment (flatcap) [reply attach/recvattach.c:560]: Yep, that's good

Comment (roccoblues) [reply attach/recvattach.c]: fixed now

**Comment by flatcap** [comment]
[I've fixed](https://github.com/neomutt/neomutt/pull/3903/commits/c40a6121f9fab920eedbf88b8b6ede5d292bf7eb) one missing leak and run [iwyu](https://neomutt.org/dev/analysis/iwyu) to tidy the includes.

**Pull request merged**